File: clients/go/mongocore/collection.go
package mongocore

import (
	"context"
	"io"

	pb "github.com/rozza/mongocore/clients/go/proto"
	"go.mongodb.org/mongo-driver/v2/bson"
)

// Collection provides CRUD operations on a MongoDB collection via MongoCore.
type Collection struct {
	client   *Client
	database string
	name     string
}

// FindOptions configures a find operation.
type FindOptions struct {
	Limit      int64
	Skip       int64
	Sort       bson.D
	Projection bson.D
}

// UpdateResult contains the result of an update operation.
type UpdateResult struct {
	MatchedCount  int64
	ModifiedCount int64
	UpsertedID    string
}

// SearchResult contains the results of a search operation.
type SearchResult struct {
	Documents []bson.D
	Method    string
	Total     int64
}

func encodeBson(doc bson.D) ([]byte, error) {
	return bson.Marshal(doc)
}

func decodeBsonDoc(data []byte) (bson.D, error) {
	var doc bson.D
	err := bson.Unmarshal(data, &doc)
	return doc, err
}

// Find returns documents matching the filter.
func (c *Collection) Find(ctx context.Context, filter bson.D, opts *FindOptions) ([]bson.D, error) {
	filterBytes, err := encodeBson(filter)
	if err != nil {
		return nil, err
	}

	req := &pb.FindRequest{
		Database:   c.database,
		Collection: c.name,
		Filter:     &pb.Filter{Data: filterBytes},
	}

	if opts != nil {
		findOpts := &pb.FindOptions{}
		if opts.Limit > 0 {
			limit := opts.Limit
			findOpts.Limit = &limit
		}
		if opts.Skip > 0 {
			skip := opts.Skip
			findOpts.Skip = &skip
		}
		if opts.Sort != nil {
			sortBytes, err := encodeBson(opts.Sort)
			if err != nil {
				return nil, err
			}
			findOpts.Sort = sortBytes
		}
		if opts.Projection != nil {
			projBytes, err := encodeBson(opts.Projection)
			if err != nil {
				return nil, err
			}
			findOpts.Projection = projBytes
		}
		req.Options = findOpts
	}

	resp, err := c.client.stub.Find(clientContext(ctx), req)
	if err != nil {
		return nil, err
	}

	docs := make([]bson.D, 0, len(resp.Documents))
	for _, d := range resp.Documents {
		doc, err := decodeBsonDoc(d.Data)
		if err != nil {
			return nil, err
		}
		docs = append(docs, doc)
	}
	return docs, nil
}

// FindOne returns a single document matching the filter.
func (c *Collection) FindOne(ctx context.Context, filter bson.D) (bson.D, error) {
	filterBytes, err := encodeBson(filter)
	if err != nil {
		return nil, err
	}

	resp, err := c.client.stub.FindOne(clientContext(ctx), &pb.FindOneRequest{
		Database:   c.database,
		Collection: c.name,
		Filter:     &pb.Filter{Data: filterBytes},
	})
	if err != nil {
		return nil, err
	}

	if resp.Document == nil || len(resp.Document.Data) == 0 {
		return nil, nil
	}
	return decodeBsonDoc(resp.Document.Data)
}

// InsertOne inserts a single document.
func (c *Collection) InsertOne(ctx context.Context, document bson.D) (string, error) {
	docBytes, err := encodeBson(document)
	if err != nil {
		return "", err
	}

	resp, err := c.client.stub.Insert(clientContext(ctx), &pb.InsertRequest{
		Database:   c.database,
		Collection: c.name,
		Document:   &pb.Document{Data: docBytes},
	})
	if err != nil {
		return "", err
	}
	return resp.InsertedId, nil
}

// InsertMany inserts multiple documents.
func (c *Collection) InsertMany(ctx context.Context, documents []bson.D) ([]string, error) {
	pbDocs := make([]*pb.Document, 0, len(documents))
	for _, doc := range documents {
		docBytes, err := encodeBson(doc)
		if err != nil {
			return nil, err
		}
		pbDocs = append(pbDocs, &pb.Document{Data: docBytes})
	}

	resp, err := c.client.stub.InsertMany(clientContext(ctx), &pb.InsertManyRequest{
		Database:   c.database,
		Collection: c.name,
		Documents:  pbDocs,
	})
	if err != nil {
		return nil, err
	}
	return resp.InsertedIds, nil
}

// UpdateOne updates a single document.
func (c *Collection) UpdateOne(ctx context.Context, filter, update bson.D) (*UpdateResult, error) {
	filterBytes, err := encodeBson(filter)
	if err != nil {
		return nil, err
	}
	updateBytes, err := encodeBson(update)
	if err != nil {
		return nil, err
	}

	resp, err := c.client.stub.Update(clientContext(ctx), &pb.UpdateRequest{
		Database:   c.database,
		Collection: c.name,
		Filter:     &pb.Filter{Data: filterBytes},
		Update:     &pb.Document{Data: updateBytes},
	})
	if err != nil {
		return nil, err
	}
	return &UpdateResult{
		MatchedCount:  resp.MatchedCount,
		ModifiedCount: resp.ModifiedCount,
	}, nil
}

// UpdateMany updates multiple documents.
func (c *Collection) UpdateMany(ctx context.Context, filter, update bson.D) (*UpdateResult, error) {
	filterBytes, err := encodeBson(filter)
	if err != nil {
		return nil, err
	}
	updateBytes, err := encodeBson(update)
	if err != nil {
		return nil, err
	}

	resp, err := c.client.stub.UpdateMany(clientContext(ctx), &pb.UpdateManyRequest{
		Database:   c.database,
		Collection: c.name,
		Filter:     &pb.Filter{Data: filterBytes},
		Update:     &pb.Document{Data: updateBytes},
	})
	if err != nil {
		return nil, err
	}
	return &UpdateResult{
		MatchedCount:  resp.MatchedCount,
		ModifiedCount: resp.ModifiedCount,
	}, nil
}

// DeleteOne deletes a single document.
func (c *Collection) DeleteOne(ctx context.Context, filter bson.D) (int64, error) {
	filterBytes, err := encodeBson(filter)
	if err != nil {
		return 0, err
	}

	resp, err := c.client.stub.Delete(clientContext(ctx), &pb.DeleteRequest{
		Database:   c.database,
		Collection: c.name,
		Filter:     &pb.Filter{Data: filterBytes},
	})
	if err != nil {
		return 0, err
	}
	return resp.DeletedCount, nil
}

// DeleteMany deletes multiple documents.
func (c *Collection) DeleteMany(ctx context.Context, filter bson.D) (int64, error) {
	filterBytes, err := encodeBson(filter)
	if err != nil {
		return 0, err
	}

	resp, err := c.client.stub.DeleteMany(clientContext(ctx), &pb.DeleteManyRequest{
		Database:   c.database,
		Collection: c.name,
		Filter:     &pb.Filter{Data: filterBytes},
	})
	if err != nil {
		return 0, err
	}
	return resp.DeletedCount, nil
}

// ChangeEvent represents a single change stream event.
type ChangeEvent struct {
	OperationType string
	Database      string
	Collection    string
	Document      bson.D
	UpdateDesc    bson.D
	DocumentKey   bson.D
}

// ChangeStream wraps a server-streaming Watch RPC. It implements io.Closer.
type ChangeStream struct {
	stream   pb.MongoCore_WatchClient
	cancelFn context.CancelFunc
}

// Next returns the next event from the change stream.
// Returns io.EOF when the stream is closed.
func (cs *ChangeStream) Next() (*ChangeEvent, error) {
	event, err := cs.stream.Recv()
	if err != nil {
		return nil, err
	}

	opNames := []string{"insert", "update", "delete", "replace", "invalidate"}
	opType := "unknown"
	if int(event.OperationType) < len(opNames) {
		opType = opNames[event.OperationType]
	}

	ce := &ChangeEvent{
		OperationType: opType,
		Database:      event.Database,
		Collection:    event.Collection,
	}

	if event.Document != nil && len(event.Document.Data) > 0 {
		doc, err := decodeBsonDoc(event.Document.Data)
		if err == nil {
			ce.Document = doc
		}
	}
	if event.UpdateDescription != nil && len(event.UpdateDescription.Data) > 0 {
		doc, err := decodeBsonDoc(event.UpdateDescription.Data)
		if err == nil {
			ce.UpdateDesc = doc
		}
	}
	if event.DocumentKey != nil && len(event.DocumentKey.Data) > 0 {
		doc, err := decodeBsonDoc(event.DocumentKey.Data)
		if err == nil {
			ce.DocumentKey = doc
		}
	}

	return ce, nil
}

// Close terminates the change stream.
func (cs *ChangeStream) Close() error {
	cs.cancelFn()
	return nil
}

// Watch opens a change stream on this collection. The returned ChangeStream must be closed when done.
func (c *Collection) Watch(ctx context.Context, pipeline []bson.D) (*ChangeStream, error) {
	stages := make([][]byte, 0, len(pipeline))
	for _, stage := range pipeline {
		stageBytes, err := encodeBson(stage)
		if err != nil {
			return nil, err
		}
		stages = append(stages, stageBytes)
	}

	watchCtx, cancel := context.WithCancel(ctx)
	stream, err := c.client.stub.Watch(clientContext(watchCtx), &pb.WatchRequest{
		Database:   c.database,
		Collection: &c.name,
		Pipeline:   &pb.Pipeline{Stages: stages},
	})
	if err != nil {
		cancel()
		return nil, err
	}

	return &ChangeStream{stream: stream, cancelFn: cancel}, nil
}

// Ensure ChangeStream implements io.Closer.
var _ io.Closer = (*ChangeStream)(nil)

// Aggregate executes an aggregation pipeline.
func (c *Collection) Aggregate(ctx context.Context, pipeline []bson.D) ([]bson.D, error) {
	stages := make([][]byte, 0, len(pipeline))
	for _, stage := range pipeline {
		stageBytes, err := encodeBson(stage)
		if err != nil {
			return nil, err
		}
		stages = append(stages, stageBytes)
	}

	resp, err := c.client.stub.Aggregate(clientContext(ctx), &pb.AggregateRequest{
		Database:   c.database,
		Collection: c.name,
		Pipeline:   &pb.Pipeline{Stages: stages},
	})
	if err != nil {
		return nil, err
	}

	docs := make([]bson.D, 0, len(resp.Documents))
	for _, d := range resp.Documents {
		doc, err := decodeBsonDoc(d.Data)
		if err != nil {
			return nil, err
		}
		docs = append(docs, doc)
	}
	return docs, nil
}

// Search performs a unified search using the best available method (vector → fulltext → filter).
func (c *Collection) Search(ctx context.Context, query string, limit int64) (*SearchResult, error) {
	resp, err := c.client.stub.Search(clientContext(ctx), &pb.SearchRequest{
		Database:   c.database,
		Collection: c.name,
		Query:      query,
		Limit:      limit,
	})
	if err != nil {
		return nil, err
	}

	docs := make([]bson.D, 0, len(resp.Documents))
	for _, d := range resp.Documents {
		doc, err := decodeBsonDoc(d.Data)
		if err != nil {
			return nil, err
		}
		docs = append(docs, doc)
	}
	return &SearchResult{
		Documents: docs,
		Method:    resp.Method,
		Total:     resp.Total,
	}, nil
}

// FindAndModify atomically modifies a document and returns it.
func (c *Collection) FindAndModify(ctx context.Context, filter, update bson.D, returnNew bool) (bson.D, error) {
	filterBytes, err := encodeBson(filter)
	if err != nil {
		return nil, err
	}
	updateBytes, err := encodeBson(update)
	if err != nil {
		return nil, err
	}

	returnDoc := pb.FindAndModifyOptions_BEFORE
	if returnNew {
		returnDoc = pb.FindAndModifyOptions_AFTER
	}

	resp, err := c.client.stub.FindAndModify(clientContext(ctx), &pb.FindAndModifyRequest{
		Database:   c.database,
		Collection: c.name,
		Filter:     &pb.Filter{Data: filterBytes},
		Update:     &pb.Document{Data: updateBytes},
		Options: &pb.FindAndModifyOptions{
			ReturnDocument: returnDoc,
		},
	})
	if err != nil {
		return nil, err
	}

	if resp.Document == nil || len(resp.Document.Data) == 0 {
		return nil, nil
	}
	return decodeBsonDoc(resp.Document.Data)
}

// CreateIndex creates an index on the collection.
func (c *Collection) CreateIndex(ctx context.Context, keys bson.D, unique bool) (string, error) {
	keysBytes, err := encodeBson(keys)
	if err != nil {
		return "", err
	}

	resp, err := c.client.stub.CreateIndex(clientContext(ctx), &pb.CreateIndexRequest{
		Database:   c.database,
		Collection: c.name,
		Keys:       &pb.Document{Data: keysBytes},
		Options: &pb.IndexOptions{
			Unique: &unique,
		},
	})
	if err != nil {
		return "", err
	}
	return resp.IndexName, nil
}

File: clients/python/src/mongocore/collection.py
"""Collection operations."""

from typing import Any, AsyncIterator, Optional
from bson import encode, decode
_CLIENT_METADATA = [("x-client-language", "python")]


class Collection:
    """A collection handle with CRUD operations."""

    def __init__(self, client, database: str, name: str):
        self._client = client
        self._database = database
        self._name = name

    def _encode_doc(self, doc: dict) -> bytes:
        """Encode a Python dict to BSON bytes."""
        return encode(doc)

    def _decode_doc(self, data: bytes) -> dict:
        """Decode BSON bytes to a Python dict."""
        return decode(data)

    def _get_stub(self):
        from .generated import mongocore_pb2_grpc
        return mongocore_pb2_grpc.MongoCoreStub(self._client.channel)

    def _make_filter(self, filter_doc: Optional[dict] = None):
        from .generated import types_pb2
        if filter_doc:
            return types_pb2.Filter(data=self._encode_doc(filter_doc))
        return types_pb2.Filter(data=b"")

    def _make_document(self, doc: dict):
        from .generated import types_pb2
        return types_pb2.Document(data=self._encode_doc(doc))

    async def find(self, filter: Optional[dict] = None, *, limit: int = 0, skip: int = 0) -> list[dict]:
        """Find documents matching the filter."""
        from .generated import mongocore_pb2, types_pb2
        stub = self._get_stub()

        options = types_pb2.FindOptions()
        if limit:
            options.limit = limit
        if skip:
            options.skip = skip

        response = await stub.Find(mongocore_pb2.FindRequest(
            database=self._database,
            collection=self._name,
            filter=self._make_filter(filter),
            options=options,
        ), metadata=_CLIENT_METADATA)
        return [self._decode_doc(doc.data) for doc in response.documents]

    async def find_one(self, filter: Optional[dict] = None) -> Optional[dict]:
        """Find a single document."""
        from .generated import mongocore_pb2
        stub = self._get_stub()
        response = await stub.FindOne(mongocore_pb2.FindOneRequest(
            database=self._database,
            collection=self._name,
            filter=self._make_filter(filter),
        ), metadata=_CLIENT_METADATA)
        if response.document and response.document.data:
            return self._decode_doc(response.document.data)
        return None

    async def insert_one(self, document: dict) -> str:
        """Insert a single document. Returns the inserted ID."""
        from .generated import mongocore_pb2
        stub = self._get_stub()
        response = await stub.Insert(mongocore_pb2.InsertRequest(
            database=self._database,
            collection=self._name,
            document=self._make_document(document),
        ), metadata=_CLIENT_METADATA)
        return response.inserted_id

    async def insert_many(self, documents: list[dict]) -> list[str]:
        """Insert multiple documents. Returns list of inserted IDs."""
        from .generated import mongocore_pb2
        stub = self._get_stub()
        response = await stub.InsertMany(mongocore_pb2.InsertManyRequest(
            database=self._database,
            collection=self._name,
            documents=[self._make_document(d) for d in documents],
        ), metadata=_CLIENT_METADATA)
        return list(response.inserted_ids)

    async def update_one(self, filter: dict, update: dict) -> dict:
        """Update a single document. Returns {matched_count, modified_count}."""
        from .generated import mongocore_pb2
        stub = self._get_stub()
        response = await stub.Update(mongocore_pb2.UpdateRequest(
            database=self._database,
            collection=self._name,
            filter=self._make_filter(filter),
            update=self._make_document(update),
        ), metadata=_CLIENT_METADATA)
        return {"matched_count": response.matched_count, "modified_count": response.modified_count}

    async def update_many(self, filter: dict, update: dict) -> dict:
        """Update multiple documents."""
        from .generated import mongocore_pb2
        stub = self._get_stub()
        response = await stub.UpdateMany(mongocore_pb2.UpdateManyRequest(
            database=self._database,
            collection=self._name,
            filter=self._make_filter(filter),
            update=self._make_document(update),
        ), metadata=_CLIENT_METADATA)
        return {"matched_count": response.matched_count, "modified_count": response.modified_count}

    async def delete_one(self, filter: dict) -> int:
        """Delete a single document. Returns deleted count."""
        from .generated import mongocore_pb2
        stub = self._get_stub()
        response = await stub.Delete(mongocore_pb2.DeleteRequest(
            database=self._database,
            collection=self._name,
            filter=self._make_filter(filter),
        ), metadata=_CLIENT_METADATA)
        return response.deleted_count

    async def delete_many(self, filter: dict) -> int:
        """Delete multiple documents. Returns deleted count."""
        from .generated import mongocore_pb2
        stub = self._get_stub()
        response = await stub.DeleteMany(mongocore_pb2.DeleteManyRequest(
            database=self._database,
            collection=self._name,
            filter=self._make_filter(filter),
        ), metadata=_CLIENT_METADATA)
        return response.deleted_count

    async def aggregate(self, pipeline: list[dict]) -> list[dict]:
        """Run an aggregation pipeline."""
        from .generated import mongocore_pb2, types_pb2
        stub = self._get_stub()

        stages = [self._encode_doc(stage) for stage in pipeline]

        response = await stub.Aggregate(mongocore_pb2.AggregateRequest(
            database=self._database,
            collection=self._name,
            pipeline=types_pb2.Pipeline(stages=stages),
        ), metadata=_CLIENT_METADATA)
        return [self._decode_doc(doc.data) for doc in response.documents]

    async def search(self, query: str, *, limit: int = 10) -> dict:
        """Search documents using the best available method (vector → fulltext → filter)."""
        from .generated import mongocore_pb2
        stub = self._get_stub()
        response = await stub.Search(mongocore_pb2.SearchRequest(
            database=self._database,
            collection=self._name,
            query=query,
            limit=limit,
        ), metadata=_CLIENT_METADATA)
        return {
            "documents": [self._decode_doc(doc.data) for doc in response.documents],
            "method": response.method,
            "total": response.total,
        }

    async def find_and_modify(self, filter: dict, update: dict, *, return_new: bool = True, upsert: bool = False) -> Optional[dict]:
        """Atomically find and modify a document, returning the result."""
        from .generated import mongocore_pb2, types_pb2
        stub = self._get_stub()
        options = types_pb2.FindAndModifyOptions(
            return_document=types_pb2.FindAndModifyOptions.AFTER if return_new else types_pb2.FindAndModifyOptions.BEFORE,
            upsert=upsert,
        )
        request = mongocore_pb2.FindAndModifyRequest(
            database=self._database,
            collection=self._name,
            filter=self._make_filter(filter),
            update=self._make_document(update),
            options=options,
        )
        response = await stub.FindAndModify(request, metadata=_CLIENT_METADATA)
        if response.document and response.document.data:
            return self._decode_doc(response.document.data)
        return None

    async def create_index(self, keys: dict, *, unique: bool = False, name: Optional[str] = None) -> str:
        """Create an index on the collection."""
        from .generated import mongocore_pb2
        stub = self._get_stub()
        request = mongocore_pb2.CreateIndexRequest(
            database=self._database,
            collection=self._name,
            keys=self._make_document(keys),
            unique=unique,
        )
        if name:
            request.name = name
        response = await stub.CreateIndex(request, metadata=_CLIENT_METADATA)
        return response.name

    def watch(self, pipeline: Optional[list[dict]] = None) -> "ChangeStream":
        """Open a change stream on this collection. Returns an async context manager."""
        return ChangeStream(self, pipeline)


class ChangeStream:
    """An async iterable change stream that auto-closes when exiting the context."""

    def __init__(self, collection: Collection, pipeline: Optional[list[dict]] = None):
        self._collection = collection
        self._pipeline = pipeline
        self._stream = None
        self._cancelled = False

    async def __aenter__(self) -> "ChangeStream":
        from .generated import mongocore_pb2, types_pb2
        stub = self._collection._get_stub()

        stages = [self._collection._encode_doc(s) for s in self._pipeline] if self._pipeline else []

        self._stream = stub.Watch(mongocore_pb2.WatchRequest(
            database=self._collection._database,
            collection=self._collection._name,
            pipeline=types_pb2.Pipeline(stages=stages) if stages else None,
        ), metadata=_CLIENT_METADATA)
        return self

    async def __aexit__(self, *exc):
        self._cancelled = True
        if self._stream is not None:
            self._stream.cancel()
        return False

    def __aiter__(self) -> AsyncIterator[dict]:
        return self

    async def __anext__(self) -> dict:
        if self._cancelled or self._stream is None:
            raise StopAsyncIteration
        try:
            event = await self._stream.read()
            if event is None:
                raise StopAsyncIteration
            result = {"operation_type": event.operation_type}
            if event.database:
                result["database"] = event.database
            if event.collection:
                result["collection"] = event.collection
            if event.document and event.document.data:
                result["document"] = self._collection._decode_doc(event.document.data)
            if event.update_description and event.update_description.data:
                result["update_description"] = self._collection._decode_doc(event.update_description.data)
            if event.document_key and event.document_key.data:
                result["document_key"] = self._collection._decode_doc(event.document_key.data)
            return result
        except Exception:
            if self._cancelled:
                raise StopAsyncIteration
            raise
